### PR TITLE
Added Fiber#cancel, which forces a Fiber to return

### PR DESCRIPTION
**Pull request by nevans**
This skips rescue and catch blocks but runs all ensure blocks. It can be run on any living fiber, and propagates cancellation to child (resumed) fibers.

Basically, I want it to behave as if you called `break` or `return` to jump all the way to the top frame of the fiber from wherever it is.. Exceptions might be too heavy-weight, can't (and shouldn't) be sent to resuming fibers, and can be caught. This can also be emulated via `catch/throw` with an anonymous `Object.new`, but that adds an extra stack frame, and implementing it reliably requires interception of `Fiber.yield`, e.g. adding a `Fiber.throw` (analogous to `Fiber.raise`) which also shouldn't be sent to resuming fibers.

I tried to match the new _(much improved!)_ transfer/resume switch semantics with cancel propagation. When the cancelation propagates all the way to the bottom of the fiber resume stack, the last fiber in the chain will then be resumed. Resuming fibers will not run until they are yielded back into.

I think it's important to allow future `resume` and `transfer` out of a canceled fiber, in case it's necessary to close resources from `ensure` blocks. However, I may have chosen the wrong behavior for multiple `cancels` on a `canceling`. I chose to use the same propagate/resume/transfer semantics but _not `break` again_ after subsequent cancels. On second thought, I think it should behave more like a standard `break`, and do exactly the same thing that `break` or `return` would normally do if they are called from an `ensure` block. I'll update this PR with that behavior when I'm able to.

This is my first big foray into the ruby C API, so I hope it makes sense what I was trying to do. Did I handle `vm_fiber_break` and `fiber_start` appropriately? Is there is somewhere more appropriate to save the cancel reason than a new `VALUE` on `rb_fiber_struct`? E.g. could I safely assign the canceled fiber's `ec->errinfo` from the calling Fiber, instead of using `cancel_reason`?

When this compiles, it warns about:
```
vm_insnhelper.h:251:1: warning: ‘vm_call_iseq_optimizable_p’ defined but not used [-Wunused-function]
  251 | vm_call_iseq_optimizable_p(const struct rb_callinfo *ci, const struct rb_callcache *cc)     
```
So I'm probably doing something wrong there, too.

### CODE REVIEW

**Comment by nevans** [comment]
I also created a ruby issue: https://bugs.ruby-lang.org/issues/17325

File: cont.c [hunk 1]
```diff
@@ -28,6 +28,9 @@
 #include "internal/scheduler.h"
 #include "mjit.h"
 #include "vm_core.h"
+#include "vm_callinfo.h"
+#include "vm_insnhelper.h"
+
```
Comment (nobu) [inline cont.c:33]: Apply https://github.com/ruby/ruby/pull/3766/checks?check_run_id=1401035342

File: cont.c [hunk 2]
```diff
@@ -1006,6 +1011,9 @@ fiber_compact(void *ptr)
 
     if (fiber->prev) rb_fiber_update_self(fiber->prev);
 
+    if (fiber->cancel_reason)
+      fiber->cancel_reason = rb_gc_location(fiber->cancel_reason);
```
Comment (nobu) [inline cont.c:1015]: Adjust the indent.

File: cont.c [hunk 3]
```diff
@@ -1927,6 +1939,18 @@ rb_fiber_start(void)
...
+    else if (state == TAG_BREAK) {
+        rb_context_t *cont = &VAR_FROM_MEMORY(fiber)->cont;
+        const struct vm_throw_data *const err = (struct vm_throw_data *)ec->errinfo;
+        const rb_control_frame_t *const escape_cfp = THROW_DATA_CATCH_FRAME(err);
+        if (cfp == escape_cfp) {
+          rb_vm_rewind_cfp(ec, cfp);
+          state = 0;
+          ec->tag->state = TAG_NONE;
+          ec->errinfo = Qnil;
+          cont->value = THROW_DATA_VAL(err);
```
Comment (nobu) [inline cont.c:1951]: Adjust the indent.

File: cont.c [hunk 4]
```diff
@@ -2101,6 +2125,8 @@ fiber_store(rb_fiber_t *next_fiber, rb_thread_t *th)
     return fiber->cont.value;
 }
 
+static void fiber_check_canceled();
```
Comment (nobu) [inline cont.c:2128]:
```suggestion
static void fiber_check_canceled(void);
```

File: cont.c [hunk 5]
```diff
@@ -2237,6 +2265,117 @@ rb_fiber_terminate(rb_fiber_t *fiber, int need_interrupt)
     fiber_switch(next_fiber, 1, &value, RB_NO_KEYWORDS, Qfalse, false);
 }
 
+static void
+fiber_check_canceled()
```
Comment (nobu) [inline cont.c:2269]:
```suggestion
fiber_check_canceled(void)
```

File: vm.c [hunk 6]
```diff
@@ -1792,6 +1792,35 @@ rb_iter_break_value(VALUE val)
     vm_iter_break(GET_EC(), val);
 }
 
+
```
Comment (nobu) [inline vm.c:1795]:
```suggestion
NORETURN(static void vm_fiber_break(rb_execution_context_t *ec, VALUE reason));

```

File: cont.c [hunk 7]
```diff
@@ -2237,6 +2265,117 @@ rb_fiber_terminate(rb_fiber_t *fiber, int need_interrupt)
...
+{
+    rb_fiber_t *fiber = fiber_current();
+    VALUE reason = fiber->cancel_reason;
+    if (fiber->canceling && reason != Qundef) {
+        VM_ASSERT(reason  != Qundef);
```
Comment (nobu) [inline cont.c:2274]: This assertion feels too paranoid, doesn't it?

File: cont.c (hunk 3 above)
Comment (nobu) [inline cont.c:1951]:
```suggestion
            rb_vm_rewind_cfp(ec, cfp);
            state = 0;
            ec->tag->state = TAG_NONE;
            ec->errinfo = Qnil;
            cont->value = THROW_DATA_VAL(err);
```

File: cont.c (hunk 2 above)
Comment (nobu) [inline cont.c:1015]:
```suggestion
    if (fiber->cancel_reason) fiber->cancel_reason = rb_gc_location(fiber->cancel_reason);
```

File: cont.c [hunk 8]
```diff
@@ -2237,6 +2265,117 @@ rb_fiber_terminate(rb_fiber_t *fiber, int need_interrupt)
...
+ *  If +#cancel+ is called more than once, only the first reason will be used.
+ *  If this fiber is resuming, a subsequent cancel will resend cancel to its
+ *  current chain of resumed fibers.
+ *
+ *  If the fiber has not been started, it will convert to terminated and
+ *  immediately return the cancelation reason. If it has already run to
+ *  completion, raises +FiberError+.
+ *
+ */
+VALUE
```
Comment (nobu) [inline cont.c:2371]:
```suggestion
static VALUE
```

File: cont.c [hunk 9]
```diff
@@ -2237,6 +2265,117 @@ rb_fiber_terminate(rb_fiber_t *fiber, int need_interrupt)
...
+ *  method will always return false.
+ */
+VALUE
+rb_fiber_canceling_p(VALUE fiber_value)
+{
+    rb_fiber_t *fiber = fiber_ptr(fiber_value);
+    return (!FIBER_TERMINATED_P(fiber) && fiber->canceling) ? Qtrue : Qfalse;
+}
+
+VALUE
```
Comment (nobu) [inline cont.c:2295]:
```suggestion
static VALUE
```

File: cont.c [hunk 10]
```diff
@@ -2237,6 +2265,117 @@ rb_fiber_terminate(rb_fiber_t *fiber, int need_interrupt)
...
+
+/*
+ *  call-seq:
+ *     fiber.canceling? -> true or false
+ *
+ *  Returns true if the fiber is has been canceled but is still alive and
+ *  executing +ensure+ blocks. After finishing execution of the fiber block this
+ *  method will always return false.
+ */
+VALUE
```
Comment (nobu) [inline cont.c:2288]:
```suggestion
static VALUE
```

File: include/ruby/internal/error.h [hunk 11]
```diff
@@ -45,6 +45,7 @@ NORETURN(void rb_mod_sys_fail_str(VALUE, VALUE));
 NORETURN(void rb_readwrite_sys_fail(enum rb_io_wait_readwrite, const char*));
 NORETURN(void rb_iter_break(void));
 NORETURN(void rb_iter_break_value(VALUE));
+NORETURN(void rb_fiber_break(VALUE));
```
Comment (nobu) [inline include/ruby/internal/error.h:48]: Should  not be in include, maybe internal/cont.h.

File: test/ruby/test_fiber.rb [hunk 12]
```diff
@@ -121,6 +121,14 @@ def test_error
     }
   end
 
+  def test_break
+    assert_raise(LocalJumpError){
+      fib = Fiber.new { break :value }
+      assert_equal(:value, fib.resume)
+      assert_equal(false, fib.alive?)
```
Comment (nobu) [inline test/ruby/test_fiber.rb:128]:
```suggestion
      assert_not_predicate(fib, :alive?)
```

File: test/ruby/test_fiber.rb [hunk 13]
```diff
@@ -169,6 +177,156 @@ def test_raise
...
+    else
+      log << :else
+    ensure
+      log << :ensured
+    end
+    assert_equal(:yield_1, fib.resume(:init))
+    assert_equal(:canceled, fib.cancel(:canceled))
+    assert_equal([:init, :ensured], log)
+    helper = CancelHelper.new
+    Fiber.new { helper.deep_call(5) }
```
Comment (nobu) [inline test/ruby/test_fiber.rb:217]: What do these lines?

File: test/ruby/test_fiber.rb [hunk 14]
```diff
@@ -169,6 +177,156 @@ def test_raise
...
+      log << Fiber.yield(:yielded_from_fiber)
+      log << :unreachable
+    ensure
+      log << Fiber.current.canceling?
+      log << Fiber.yield(:yielded_from_ensure1)
+      log << Fiber.current.canceling?
+      log << Fiber.yield(:yielded_from_ensure2)
+    end
+    log << f.resume
+    assert(!f.canceling?)
```
Comment (nobu) [inline test/ruby/test_fiber.rb:261]:
```suggestion
    assert_not_predicate(f, :canceling?)
```

File: test/ruby/test_fiber.rb [hunk 15]
```diff
@@ -169,6 +177,156 @@ def test_raise
...
+    log << f.cancel(:canceled)
+    assert(f.canceling?)
```
Comment (nobu) [inline test/ruby/test_fiber.rb:263]:
```suggestion
    assert_predicate(f, :canceling?)
```

File: test/ruby/test_fiber.rb [hunk 16]
```diff
@@ -169,6 +177,156 @@ def test_raise
...
+    log << f.resume(:continue1)
+    assert(f.canceling?)
```
Comment (nobu) [inline test/ruby/test_fiber.rb:265]:
```suggestion
    assert_predicate(f, :canceling?)
```

File: test/ruby/test_fiber.rb [hunk 17]
```diff
@@ -169,6 +177,156 @@ def test_raise
...
+    log << f.resume(:continue2)
+    assert_equal([
+      false, :yielded_from_fiber, # cancel called
+      true, :yielded_from_ensure1, :continue1,
+      true, :yielded_from_ensure2, :continue2,
+      :canceled,
+    ], log)
+    assert(!f.canceling?)
```
Comment (nobu) [inline test/ruby/test_fiber.rb:273]:
```suggestion
    assert_not_predicate(f, :canceling?)
```

File: test/ruby/test_fiber.rb [hunk 18]
```diff
@@ -169,6 +177,156 @@ def test_raise
...
+    assert_equal([:txfr1_from_root, :txfr2_from1, :ensure1], log)
+    log.clear
+    assert_equal(:return2, f2.transfer(:txfr2_from_root))
+    assert_equal([:txfr2_from_root, :done2, :ensure2], log)
+  end
+
+  def test_cancel_unborn_fiber
+    fib = Fiber.new{ :unreachable }
+    assert_equal("cancel unborn", fib.cancel("cancel unborn"))
+    assert(!fib.alive?)
```
Comment (nobu) [inline test/ruby/test_fiber.rb:308]:
```suggestion
    assert_not_predicate(fib, :alive?)
```

File: test/ruby/test_fiber.rb [hunk 19]
```diff
@@ -169,6 +177,156 @@ def test_raise
...
+      fib = Fiber.new{}
+      fib.resume
+      fib.cancel "cancel in dead fiber"
+    }
+  end
+
+  def test_cancel_unborn_and_dead_fibers
+    fib = Fiber.new{ :unreachable }
+    assert_equal("cancel unborn", fib.cancel("cancel unborn"))
+    assert(!fib.alive?)
```
Comment (nobu) [inline test/ruby/test_fiber.rb:322]:
```suggestion
    assert_not_predicate(fib, :alive?)
```

File: cont.c (hunk 7 above)
Comment (nevans) [reply cont.c:2274]: Haha, yes. it got left in from an earlier revision. :)

File: cont.c (hunk 1 above)
Comment (nevans) [reply cont.c:33]: Oh, thanks! When I run `make all golf` followed by `ruby tool/update-deps --fix` locally, it doesn't change anything. But I applied the diff manually.

File: test/ruby/test_fiber.rb (hunk 13 above)
Comment (nevans) [reply test/ruby/test_fiber.rb:217]: oops. I didn't finish that test!!

**Comment by nevans** [comment]
@nobu thank you _so much_ for your _very_ helpful review, spotting and cleaning up my many mistakes. I'm fixing all of your suggestions now. I haven't written C in over 20 years, and I felt a bit lost some times! :slightly_smiling_face: 

I'm still seeing this warning, is that okay?
```
In file included from cont.c:32:
vm_insnhelper.h:251:1: warning: ‘vm_call_iseq_optimizable_p’ defined but not used [-Wunused-function]
  251 | vm_call_iseq_optimizable_p(const struct rb_callinfo *ci, const struct rb_callcache *cc)     
```

**Comment by nevans** [comment]
Like I wrote in the description, I'm also going to update the code so that multiple calls to `cancel` while it is in `ensure` behaves the same way as calling `break` or `return` from inside an `ensure` block. I think that will be a lot less confusing and less error prone. As it is currently implemented, it breaks the symmetry of "transferring out will return the value from a transfer back in" and "yielding out will return the value from a resume back in". As currently implemented, calling `fiber.cancel(reason)` will result in `Fiber.yield` inside the fiber returning a `reason`, which will almost certainly be a bug. `fiber.cancel` should always `break` in the same way that `fiber.raise` always raises an exception.

As a result of that, multiple calls to `cancel` will result in the _last_ reason being the one that is returned, instead of the first reason. I think that's probably okay.

**Comment by nobu** [comment]
> I'm still seeing this warning, is that okay?
> 
> ```
> 
> In file included from cont.c:32:
> 
> vm_insnhelper.h:251:1: warning: ‘vm_call_iseq_optimizable_p’ defined but not used [-Wunused-function]
> 
>   251 | vm_call_iseq_optimizable_p(const struct rb_callinfo *ci, const struct rb_callcache *cc)     
> 
> ```

Added `inline` to that definition.
It should disappear by rebasing.

**Comment by nevans** [comment]
@nobu thank you so much for your help on this. I know it'll need to be discussed and might not be accepted, so thanks again for spending the time to help me clean it up.

I did rebase and that warning did go away.

I also updated https://bugs.ruby-lang.org/issues/17325 with a (too verbose) explanation for why I think that something like this is an important feature, even with `Fiber#raise` and even if my implementation isn't used.

I think I'd like to change my PR so that `terminated_fiber.cancel` returns `nil` or `false` instead of raising `FiberError`. That way any fiber can be canceled without exceptions (unless an exception is raised during cancellation).

This PR is for non-exceptional cancellation (doesn't raise in `return_fiber`) that are not catchable (like `break`). It might be worthwhile to add keywords options for all four possibilities:

| use break or raise to cancel | on completion, raise error in return_fiber | method call | n.b. |
|-|-|-|-|
| break (unrescuable) | unexceptional | `fiber.cancel` | _this PR. the common case_ |
| break (unrescuable) | exceptional | `fiber.cancel(raise: err)` | _exceptional cancellation, e.g. most timeouts_ |
| raise (rescuable) | unexceptional | `fiber.cancel(with_exception: err)` | _preventable cancellation_ |
| raise (rescuable) | exceptional | `fiber.cancel(with_exception: err, raise: err)` | _Fiber#raise with "resuming" propagation_ |

I think the first one is most useful. The others could be in a future PR, if this one (or something like it) is accepted.

**Comment by nevans** [comment]
I'm closing this PR.  But I would like ruby to have this (and other) concurrency primitives at some point, using `Fiber.current.storage` (#6612) and/or FiberScheduler.

**Pull request closed without merging**